src/compiler/tsbuild.ts

```diff
@@ -1182,6 +1183,97 @@ namespace ts {
+            const readFileCache = createMap<string | false>();
+            const fileExistsCache = createMap<boolean>();
+            const directoryExistsCache = createMap<boolean>();
+            const sourceFileCache = createMap<SourceFile>();
+            const savedReadFileWithCache = readFileWithCache;
+
+            // TODO:: In watch mode as well to use caches for incremental build once we can invalidate caches correctly and have right api
+            // Override readFile for json files and output .d.ts to cache the text
+            readFileWithCache = fileName => {
+                const key = toPath(fileName);
+                const value = readFileCache.get(key);
+                if (value !== undefined) return value || undefined;
+                return setReadFileCache(key, fileName);
+            };
+            const setReadFileCache = (key: Path, fileName: string) => {
+                const newValue = originalReadFile.call(host, fileName);
+                readFileCache.set(key, newValue || false);
+                return newValue;
+            };
+            host.readFile = fileName => {
+                const key = toPath(fileName);
+                const value = readFileCache.get(key);
+                if (value !== undefined) return value; // could be .d.ts from output
+                if (!fileExtensionIs(fileName, Extension.Json)) {
+                    return originalReadFile.call(host, fileName);
+                }
+
+                return setReadFileCache(key, fileName);
+            };
+            host.getSourceFile = (fileName, languageVersion, onError, shouldCreateNewSourceFile) => {
+                const key = toPath(fileName);
+                const value = sourceFileCache.get(key);
+                if (value) return value;
+
+                const sourceFile = originalGetSourceFile.call(host, fileName, languageVersion, onError, shouldCreateNewSourceFile);
+                if (sourceFile && (isDeclarationFileName(fileName) || fileExtensionIs(fileName, Extension.Json))) {
+                    sourceFileCache.set(key, sourceFile);
+                }
+                return sourceFile;
+            };
+
+            // fileExits for any kind of extension
+            host.fileExists = fileName => {
+                const key = toPath(fileName);
+                const value = fileExistsCache.get(key);
+                if (value !== undefined) return value;
+                const newValue = originalFileExists.call(host, fileName);
+                fileExistsCache.set(key, !!newValue);
+                return newValue;
+            };
+            host.writeFile = (fileName, data, writeByteOrderMark, onError, sourceFiles) => {
+                const key = toPath(fileName);
+                fileExistsCache.delete(key);
+
+                const value = readFileCache.get(key);
+                if (value && value !== data) {
+                    readFileCache.delete(key);

```

Comment:
can you update the caches here instead?

Reply:
Deleting fileExists cache key since (writeFile method can fail? say because of write permisions and so on)
Same for sourceFileCache (also we shouldnt need to create it if its not required later)